File: pawn.py
from typing import List, Set
from constants import PieceType, Color, Direction
from move import Move
from piece import Piece
from square import Square

class Pawn (Piece):
    def __init__(self, color: Color, square: Square):
        super().__init__(PieceType.PAWN, color, square)
        self.isLight = False

        # TODO - these numbers are the same for all pawns of the same player, they should be moved somewhere
        if self.color == Color.WHITE:
            self.MOVE_OFFSET = 1  # white pawns move up the board (+1 row), black opposite
            self.BASE_ROW = 1  # white pawns start on row 1, black on row 6
            self.PROMOTION_ROW = 7  # white pawns promote on row 7, black on row 0
            self.EN_PASSANT_ROW = 4  # white pawns can take en passant on row 4, black on row 3
        else:
            self.MOVE_OFFSET = -1
            self.BASE_ROW = 6
            self.PROMOTION_ROW = 0
            self.EN_PASSANT_ROW = 3

    def calc_potential_moves(self) -> List[Move]:
        """
        potential moves of a piece are not necessarily legal moves, checks and pins are not considered
        the potential moves consists of possible (1) moves forward, (2) captures left, (3) captures right and (4) en passant
        """
        return self.get_forward_moves() + self.get_capture_moves(1) + self.get_capture_moves(-1) + self.get_en_passant_moves()

    def get_en_passant_moves(self) -> List[Move]:
        """check the possibility of taking en passant"""

        potentialMoves: List[Move] = []
        enPassantSqr = self.square.board.enPassantPawnSquare[-1] # if en passant is possible, the enPassantSqr is the square behind the pawn

        # if en passant is possible and the pawn to take is next to this pawn, add the en passant move
        if (enPassantSqr is not None and self.square.rowIdx == self.EN_PASSANT_ROW and abs(self.square.idx - enPassantSqr.idx) == 1 and
                # en passant pin is a bit problematic, because the pawn is not technically pinned so this case cannot be handled in
                # "calc_potential_moves_pinned" and there is no suitable place to handle it
                # the en passant move should technically be among potential moves even if the pawn is pinned in this way, but if it's not,
                # it shouldn't break anything
                not self.is_en_passant_pin(enPassantSqr)):

            toSqr = self.square.board.get_square_by_coords(enPassantSqr.colIdx, enPassantSqr.rowIdx + self.MOVE_OFFSET)
            return [Move(self, toSqr, isEnPassant=True)]

        return []

    def get_forward_moves(self) -> List[Move]:
        """check the possibility of moving one step forward or two steps forward if on the base row"""

        potentialMoves: List[Move] = []
        # check the square in front of the pawn
        square = self.square.board.get_square_by_coords(self.square.colIdx, self.square.rowIdx + self.MOVE_OFFSET)
        if square.piece is None:
            potentialMoves += self.generate_pawn_move(square) # might be a promotion - that would be 4 possible moves

            # if on the base row, check one square further
            if self.square.rowIdx == self.BASE_ROW:
                square = self.square.board.get_square_by_coords(self.square.colIdx, self.square.rowIdx + 2 * self.MOVE_OFFSET)
                if square.piece is None:
                    potentialMoves.append(Move(self, square))
        return potentialMoves

    def get_capture_moves(self, columnOffset: int) -> List[Move]:
        """check the possibility of capturing a piece on the given column offset: left(-1) or right(1)"""

        square = self.square.board.get_square_by_coords_opt(self.square.colIdx + columnOffset, self.square.rowIdx + self.MOVE_OFFSET)
        if square is not None and square.piece is not None and square.piece.color != self.color:
            return self.generate_pawn_move(square)
        return []

    def calc_potential_moves_pinned(self, direction: Direction) -> List[Move]:
        """the piece is pinned in the given direction, it can potentially still move in the pin and the opposite direction"""

        potentialMoves: List[Move] = []
        if direction == Direction.RIGHT or direction == Direction.LEFT:
            # a pawn pinned from side can never move
            return potentialMoves

        elif direction == Direction.UP or direction == Direction.DOWN:
            # a pawn pinned from front or back can only move forward
            potentialMoves += self.get_forward_moves()

        elif direction == Direction.UP_RIGHT or direction == Direction.DOWN_LEFT:
            # pawn pinned diagonally can possibly capture in the pin direction
            potentialMoves += self.get_capture_moves(self.MOVE_OFFSET)

            # a diagonally pinned pawn can even capture en passant, en passant pin special pin is impossible in this case (the pawn is already pinned)
            enPassant = self.square.board.enPassantPawnSquare[-1]
            if enPassant is not None and self.square.rowIdx == self.EN_PASSANT_ROW and enPassant.idx - self.square.idx == self.MOVE_OFFSET:
                potentialMoves.append(Move(self, self.square.board.get_square_by_coords(enPassant.colIdx, enPassant.rowIdx + self.MOVE_OFFSET),
                                           isEnPassant=True))

        else: # direction is LEFT_UP or RIGHT_DOWN
            # capture in the other direction is analogous to the previous case
            potentialMoves += self.get_capture_moves(-self.MOVE_OFFSET)
            enPassant = self.square.board.enPassantPawnSquare[-1]
            if enPassant is not None and self.square.rowIdx == self.EN_PASSANT_ROW and self.square.idx - enPassant.idx == self.MOVE_OFFSET:
                potentialMoves.append(Move(self, self.square.board.get_square_by_coords(enPassant.colIdx, enPassant.rowIdx + self.MOVE_OFFSET),
                                           isEnPassant=True))

        return potentialMoves

    def generate_pawn_move(self, targetSquare: Square) -> List[Move]:
        """normally the pawn has one possible move to the given square, but if it's a promotion, there are 4 possible moves"""

        if targetSquare.rowIdx != self.PROMOTION_ROW:
            return [Move(self, targetSquare)]
        else:
            return [Move(self, targetSquare, newPiece=piece) for piece in [PieceType.KNIGHT, PieceType.BISHOP, PieceType.ROOK,
                                                                                             PieceType.QUEEN]]
    
    def calc_attacked_squares(self) -> Set[Square]:
        """calculates squares attacked by the piece"""

        if self.square.colIdx == 0:
            return {self.square.board.get_square_by_coords(self.square.colIdx + 1, self.square.rowIdx + self.MOVE_OFFSET)}
        elif self.square.colIdx == 7:
            return {self.square.board.get_square_by_coords(self.square.colIdx - 1, self.square.rowIdx + self.MOVE_OFFSET)}
        else:
            return {
                self.square.board.get_square_by_coords(self.square.colIdx + 1, self.square.rowIdx + self.MOVE_OFFSET),
                self.square.board.get_square_by_coords(self.square.colIdx - 1, self.square.rowIdx + self.MOVE_OFFSET)
            }
        
    def is_en_passant_pin(self, enPassantSquare: Square) -> bool:
        """
        checks very specific situation when the pawn is not technically pinned - there is another pawn in the path, but taking en passant would
        still expose the king to check (like white: Kh5, pf5, black Qe5, pg5, white to move, taking en passant would expose the king to check,
        even though the f5 pawn is technically not pinned)
        """
        assert enPassantSquare is not None and enPassantSquare.piece is not None and enPassantSquare.piece.kind == PieceType.PAWN and\
            enPassantSquare.piece.color != self.color, f"no opponent's pawn to take en passant"
        king = self.square.board.get_king(self.color)
        if not king: # this cannot occur in a regular game
            return False

        # the special pin can only occur on the en passant row
        if  king.square.rowIdx != enPassantSquare.rowIdx:
            return False

        # what is the direction of the potential pin (from king to pinner)
        direction = Direction.LEFT if king.square.colIdx > enPassantSquare.colIdx else Direction.RIGHT

        # on the way from own king towards the pinner, the first piece must be the own pawn (that wants to take en passant) or the opponent's pawn
        firstSquare = self.square.board.find_first_occupied_square_in_dir(king.square, direction)
        if firstSquare is None or (firstSquare != enPassantSquare and firstSquare.piece != self):
            return False

        # second piece must be the other relevant pawn (taker or being taken) and must be right next to the first piece
        secondSquare = self.square.board.get_square_by_coords_opt(firstSquare.colIdx + (1 if direction == Direction.RIGHT else -1), firstSquare.rowIdx)
        if secondSquare is None or secondSquare.piece is None:
            return False

        # now check that those two pawns have been found
        if not (firstSquare == enPassantSquare and secondSquare.piece == self or firstSquare.piece == self and secondSquare == enPassantSquare):
            return False

        # the third piece must be the opponent's queen or rook
        thirdSquare = self.square.board.find_first_occupied_square_in_dir(secondSquare, direction)
        if thirdSquare is None:
            return False
        assert thirdSquare.piece is not None, f"find_first_occupied_square_in_dir returned an empty square"
        if thirdSquare.piece.color == self.color or thirdSquare.piece.kind not in [PieceType.QUEEN, PieceType.ROOK]:
            return False

        # all conditions are met, the pawn is pinned
        return True

    def __str__(self):
        return f'p{self.square.name}'

    def __repr__(self):
        return f'Pawn({self.color}, {self.square})'

File: c_engine/src/pawn.cpp
#include "pawn.h"
#include "board.h"
#include "constants.h"
#include "move.h"

 Pawn::Pawn(Color color, Square *square)
    : Piece(PieceType::PAWN, color, square),
    _moveOffset(color == Color::WHITE ? 1 : -1),
    _baseRow(color == Color::WHITE ? 1 : 6),
    _promotionRow(color == Color::WHITE ? 7 : 0),
    _enPassantRow(color == Color::WHITE ? 4 : 3) {

    _name = "p";
    _isSliding = false;
    _isLight = false;
    _attackedSquares.reserve(2);
}

std::vector<Square *> Pawn::get_potential_squares() {
    std::vector<Square *> squares;

    for (auto square : get_forward_squares()) // move forward
        squares.push_back(square);

    for (auto square : get_capture_squares(-1)) // capture to the left
        squares.push_back(square);

    for (auto square : get_capture_squares(1)) // capture to the right
        squares.push_back(square);

    auto enPassantSquare = get_en_passant_square(); // en passant
    if (enPassantSquare != nullptr)
        squares.push_back(enPassantSquare);

    return squares;
}

Square * Pawn::get_en_passant_square() {
    // check the possibility that the pawn can take en passant
    // if en passant is possible, the enPassantSqr is the square behind the pawn
    auto enPassantSqr = _square->get_board()->get_en_passant_pawn_square();
    if (enPassantSqr == nullptr)
        return nullptr;

    // the pawn must be on the right row (4 for white, 3 for black indexed from 0)
    if (_square->get_row() != _enPassantRow)
        return nullptr;

    // the pawn must be on the neighboring column of the en passant square
    if (abs(_square->get_col() - enPassantSqr->get_col()) != 1)
        return nullptr;

    if (is_en_passant_pin())
        return nullptr;

    return _square->get_board()->get_square(enPassantSqr->get_col(), _square->get_row() + _moveOffset);
}

bool Pawn::is_en_passant_pin() {
    // handles a very special case, in which the pawn is technically not pinned, but en passant capture would
    // still expose the king
    // e.g. white:  pe5, Kg5, black: Rb5, pd7 and black moves pd7-d5 - then "exd6 e.p."" is not possible 

    auto enPassantSqr = _square->get_board()->get_en_passant_pawn_square();
    ASSERT(enPassantSqr != nullptr, "en passant square is null");
    ASSERT(enPassantSqr->get_piece() != nullptr, "en passant square is free");
    ASSERT(enPassantSqr->get_piece()->_kind == PieceType::PAWN, "en passant square is not occupied by a pawn");
    ASSERT(enPassantSqr->get_piece()->_color != _color, "en passant square is occupied by a pawn of the same color");

    auto king = _square->get_board()->get_king(_color);
    if (king == nullptr)  // cannot happen in a regular game
        return false;

    // the special pin can only occur on the en passant row
    if (king->get_square()->get_row() != enPassantSqr->get_row())
        return false;

    // what is the direction of the potential pin (from king to pinner)
    auto direction = king->get_square()->get_col() > enPassantSqr->get_col() ? Direction::LEFT : Direction::RIGHT;

    // on the way from own king towards the pinner, the first piece must be the own pawn (that wants to take en passant)
    // or the opponent's pawn
    auto firstSquare = _square->get_board()->find_first_occupied_square_in_dir(king->get_square(), direction);
    if (firstSquare == nullptr || (firstSquare != enPassantSqr && firstSquare->get_piece() != this))
        return false;

    // the second piece must be the other pawn (taker or being taken) and must be right next to the first piece
    auto secondSquare = _square->get_board()->
        get_square(firstSquare->get_col() + (direction == Direction::RIGHT ? 1 : -1), firstSquare->get_row());
    if (secondSquare == nullptr || secondSquare->get_piece() == nullptr)
        return false;

    // now check that those two pawns have been found
    bool firstSquareIsEnPassant = firstSquare == enPassantSqr && secondSquare->get_piece() == this;
    bool secondSquareIsEnPassant = firstSquare->get_piece() == this && secondSquare == enPassantSqr;
    if (!(firstSquareIsEnPassant || secondSquareIsEnPassant))
        return false;


    // the third piece must be the opponent's queen or rook
    auto thirdSquare = _square->get_board()->find_first_occupied_square_in_dir(secondSquare, direction);
    if (thirdSquare == nullptr)
        return false;

    // the third piece must be the opponent's queen or rook
    if (thirdSquare->get_piece()->_color == _color)
        return false;
    if (thirdSquare->get_piece()->_kind != PieceType::QUEEN && thirdSquare->get_piece()->_kind != PieceType::ROOK)
        return false;

    // all conditions are met, the pawn is "pinned"
    return true;
}

std::vector<Square *> Pawn::get_capture_squares(int colOffset) {
    std::vector<Square *> squares;

    auto sqr = _square->get_board()->get_square(_square->get_coordinate() + Coordinate{colOffset, _moveOffset});
    if (sqr != nullptr && !sqr->is_free() && sqr->get_piece()->_color != _color)
        squares.push_back(sqr);
        //for (auto move : generate_pawn_moves(sqr))
        //    moves.push_back(move);

    return squares;
}

std::vector<Square *> Pawn::get_forward_squares() {
    std::vector<Square *> squares;

    // if the square in front if the pawn is free, it can move there
    auto sqr = _square->get_board()->get_square(_square->get_coordinate() + Coordinate{0, _moveOffset});
    ASSERT(sqr != nullptr, "square in front of pawn is not valid");
    if (sqr->is_free()) {
        //for (auto square : generate_pawn_moves(sqr))
        squares.push_back(sqr);

        // if the pawn is on its base row, it can move two squares forward
        if (_square->get_coordinate().row == _baseRow) {
            sqr = _square->get_board()->get_square(_square->get_coordinate() + Coordinate{0, 2 * _moveOffset});
            ASSERT(sqr != nullptr, "square two squares in front of pawn is not valid");
            if (sqr->is_free())
                //for (auto move : generate_pawn_moves(sqr))
                    squares.push_back(sqr);
        }      
    }

    return squares;
}

void Pawn::recalculate() {
    // pawn doesn't store potential moves, only attacked squares (possible todo)

    for (auto sqr : _attackedSquares) {
        // QQQ
        // auto& attackedBy = sqr->get_attacked_by(_color);
        // attackedBy.erase(std::find(attackedBy.begin(), attackedBy.end(), this));
        auto attackedBy = &sqr->get_attacked_by(_color);
        attackedBy->erase(std::find(attackedBy->begin(), attackedBy->end(), this));

    }
    _attackedSquares.clear();

    if (_square->get_col() == 0)
        _attackedSquares.push_back(_square->get_board()->get_square(1, _square->get_row() + _moveOffset));
    else if (_square->get_col() == 7)
        _attackedSquares.push_back(_square->get_board()->get_square(6, _square->get_row() + _moveOffset));
    else {
        _attackedSquares.push_back(_square->get_board()->get_square(_square->get_col() - 1, _square->get_row() + _moveOffset));
        _attackedSquares.push_back(_square->get_board()->get_square(_square->get_col() + 1, _square->get_row() + _moveOffset));
    }

    for (auto sqr : _attackedSquares)
        sqr->get_attacked_by(_color).push_back(this);
}

std::vector<Square *> Pawn::calc_potential_squares_pinned(Direction directionFromKingToPinner) {
    // the piece is pinned in the given direction, it can potentially still move in the pin and the opposite direction

    if (directionFromKingToPinner == Direction::RIGHT || directionFromKingToPinner == Direction::LEFT)
        // a pawn pinned from side can never move
        return {};

    if (directionFromKingToPinner == Direction::UP || directionFromKingToPinner == Direction::DOWN)
        // a pawn pinned from front or back can only move forward
        return get_forward_squares();

    auto enPassantSqr = _square->get_board()->get_en_passant_pawn_square();
    if (directionFromKingToPinner == Direction::UP_RIGHT || directionFromKingToPinner == Direction::DOWN_LEFT) {
        // pawn pinned diagonally can possibly capture in the pin direction
        std::vector<Square *> squares = get_capture_squares(_moveOffset);

        // a diagonally pinned pawn can even capture en passant, en passant pin special pin is impossible in this case
        // as the pawn is already pinned
        if (enPassantSqr != nullptr && _square->get_row() == _enPassantRow && enPassantSqr->get_idx() - _square->get_idx() == _moveOffset) {
            squares.push_back(_square->get_board()->get_square(enPassantSqr->get_col(), _square->get_row() + _moveOffset));
            //moves.back()->mark_as_en_passant();
        }

        return squares;
    }

    // directionFromKingToPinner == Direction::UP_LEFT || directionFromKingToPinner == Direction::DOWN_RIGHT
    // capture in the other direction is analogous to the previous case
    std::vector<Square *> squares = get_capture_squares(-_moveOffset);

    if (enPassantSqr != nullptr && _square->get_row() == _enPassantRow && _square->get_idx() - enPassantSqr->get_idx() == _moveOffset) {
        squares.push_back(_square->get_board()->get_square(enPassantSqr->get_col(), _square->get_row() + _moveOffset));
        //moves.back()->mark_as_en_passant();
    }

    return squares;
}